specification/k3po.control/SPEC.md

```diff
@@ -0,0 +1,258 @@
+response - STARTED when the execution of the scripts is started and FINISHED when the execution of the scripts is finished.
+
+Throughout the flow description bellow, **an error** represents an ERROR event sent by the driver.
+
+- The test framework initiates the conversation with a **PREPARE** command. This command will make the k3po server load and prepare a set
+of scripts for execution. The scripts are not yet executed, just loaded from disk and parsed. If there are ACCEPT instructions in the scripts,
+a bound on corresponding ports is performed. If the command is successful, the driver sends back a PREPARED event. In case of failure, an
+ERROR event is sent.
+
+- If the PREPARED command is successful, a **START** command may be sent by the test framework. This command will start the execution of the
+ clients in the scripts that were sent in the PREPARE command. If the command is successful, the driver sends back a STARTED event. This 
+ command will also trigger the sending of a FINISHED event whenever the scripts finished execution on the k3po server.
+ 
+- If there is no PREPARE command send before the START command or if the PREPARE command was not successful an error with a message "Script has not
+ been prepared or is still preparing" must occur. 
+ 
+- If the START command was already sent, an error with a message "Script has already been started" must occur.
+
+- Any time after the PREPARE command and before the DISPOSE command, the test framework can send an **AWAIT** command for a barrier existing 
+in the scripts. The command receives a response with a NOTIFIED event only when the corresponding barrier is notified. In case the barrier 
+does not exist, an error must be returned.
+
+- Any time after the PREPARE command *and before DISPOSE?*, the test framework can send a **NOTIFY** command for a barrier existing in the scripts.
+ The command notifies the barrier and a response with a NOTIFIED event is received. In case the barrier does not exist, an error must be returned.
+
+- Any time after the PREPARE command *and before DISPOSE?*, the test framework can send an **ABORT** command. The driver must abort the current 
+execution. It may be sent any time after a PREPARE command.
+
+- Any time after the PREPARE command, the test framework can send a **DISPOSE** command. The driver must abort the execution and release the resources 
+in use. When done, a DISPOSED event must be sent back.
+
+
+TODO - after the test is finished, you can start a new test from the beginning, with a new PREPARE. This does not seem to work very well. Usually 
+the server gets stuck. Sometimes even a new connection will not help, still the test will not work. Should we support this or should we force the 
+closing of the connection and opening a new one ? (refuse any command after a DISPOSE)
+
+
+TODO - at this moment, nothing is returned by the driver after an ABORT command. I think we should return something. A FINISHED is not acceptable,
```

Comment:
What benefit or edge cases do you think this would bring to the lifecycle of the driver or the test framework? 

From the test frameworks point of view, an abort is typically sent if the test times out.  This indicates that the test is not making progress on the executed script.  The driver receives the abort, and then sends a finish back with the progress that was made.   This way the test framework can say something like test failed due to script mismatched exception (shows diff), caused by test timeout.